src/array-repeat-strategy.js

```diff
@@ -20,16 +20,82 @@ export class ArrayRepeatStrategy {
   * @param items The new array instance.
   */
   instanceChanged(repeat, items) {
-    if (repeat.viewsRequireLifecycle) {
-      let removePromise = repeat.removeAllViews(true);
-      if (removePromise instanceof Promise) {
-        removePromise.then(() => this._standardProcessInstanceChanged(repeat, items));
+    const itemNameInBindingContext = repeat.local;
+    
+    const children = repeat.viewSlot.children;
+    let itemsPreviouslyInViews = children.map(view => view.bindingContext[itemNameInBindingContext]);
+    
+    let itemsLength = items.length;
+    
+    // remove items no longer in the instance
+    if (itemsPreviouslyInViews.length > 0) {
+      let indiciesToRemove = [];
+      
+      itemsPreviouslyInViews.forEach((oldItem, index) => {
+        if (items.indexOf(oldItem) === -1) {
+          indiciesToRemove.push(index);
+        }
+      })
+      let rmPromises = [];
+      
+      for (let i = indiciesToRemove.length - 1; i >= 0; i--) {
+        let viewOrPromise = repeat.removeView(indiciesToRemove[i], true);
+        if (viewOrPromise instanceof Promise) {
+          rmPromises.push(viewOrPromise);
+        }
+        
+        // remove also from the cache list:
+        itemsPreviouslyInViews.splice(indiciesToRemove[i], 1);
+      }
+      
+      // update views
+      Promise.all(rmPromises).then(() => {
+        let index = 0;
+        for (let item of items) {
+          let indexOfView = itemsPreviouslyInViews.indexOf(item);
+          let view;
+          
+          if (indexOfView === -1) { // create views for new items
+            let overrideContext = createFullOverrideContext(repeat, items[index], index, itemsLength);
+            repeat.insertView(index, overrideContext.bindingContext, overrideContext);
+            
+            // reflect the change in our cache list so indicies are valid
+            itemsPreviouslyInViews.splice(index, 0, undefined);
+          }
+          else if (indexOfView === index) { // leave unchanged items
+            view = children[indexOfView];
+            itemsPreviouslyInViews[indexOfView] = undefined;
+          }
+          else { // move the element to the right place
+            view = children[indexOfView];
+            repeat.moveView(indexOfView, index);
+            itemsPreviouslyInViews.splice(indexOfView, 1);
+            itemsPreviouslyInViews.splice(index, 0, undefined);
+          }
+          
+          if (view) {
+            updateOverrideContext(view.overrideContext, index, itemsLength);
```

Comment:
Indeed! I've had it, but forgot to add that in the cleaned up version (it's on the gist.run demo though).